### PR TITLE
"Save As..." for exporting copies of Realm files

### PR DESCRIPTION
**Pull request by TimOliver**

### CODE REVIEW

File: RealmBrowser/Base.lproj/MainMenu.xib [hunk 1]
```diff
@@ -97,6 +97,11 @@
                                 </menu>
                             </menuItem>
                             <menuItem isSeparatorItem="YES" id="WZk-W9-hhB"/>
+                            <menuItem title="Save As..." keyEquivalent="S" id="hoM-YX-5ZD">
```
Comment (jpsim) [inline RealmBrowser/Base.lproj/MainMenu.xib]: "command-shift-s" :+1:

File: RealmBrowser/Classes/RLMRealmBrowserWindowController.m [hunk 2]
```diff
@@ -101,6 +101,20 @@ - (IBAction)saveObjcModels:(id)sender
...
+- (IBAction)saveCopy:(id)sender
+{
+    NSString *fileName = [self.modelDocument.presentedRealm.realm.path lastPathComponent];
+    NSSavePanel *panel = [NSSavePanel savePanel];
+    [panel setNameFieldStringValue:fileName];
+    [panel beginSheetModalForWindow:self.window completionHandler:^(NSInteger result){
+        if (result == NSFileHandlingPanelOKButton)
+        {
+            NSURL *fileURL = [panel URL];
+            [self.modelDocument.presentedRealm.realm writeCopyToPath:fileURL.path error:nil];
```
Comment (jpsim) [inline RealmBrowser/Classes/RLMRealmBrowserWindowController.m]: This should handle errors, and ideally also run `optimize()` on the tables (see [`-[RLMRealm compact]`](https://github.com/realm/realm-cocoa/blob/v0.94.1/Realm/RLMRealm.mm#L634-L674) for how to do this).

Comment (TimOliver) [inline RealmBrowser/Classes/RLMRealmBrowserWindowController.m]: Fair enough! I've added error handling now.

For now, I just exposed `[RLMRealm compact]` and had that run on the new Realm file. Are there any downsides to that?

Comment (bdash) [inline RealmBrowser/Classes/RLMRealmBrowserWindowController.m]: Is doing `-compact` implicitly a good idea? The user may just want to save a copy of their Realm with free space removed, not with the other changes that `-compact` makes to the structure.

Comment (jpsim) [inline RealmBrowser/Classes/RLMRealmBrowserWindowController.m]: The changes that `-compact` makes to the underlying file structure are just implementation details that are largely invisible to the end user, and an improvement in most situations. I think it's fine to generalize and optimize for the common use case in a tool like this.

**Comment by jpsim** [comment]
why is this still WIP, @TimOliver ?

**Comment by TimOliver** [comment]
@jpsim Sorry! I was watching this, but I wasn't sure if we were all happy with using the `compact` method.

Are we all good? Can I merge this?

**Comment by jpsim** [comment]
I think this functionality needs to be accompanied by a progress indicator, as copying & compacting the file could potentially take a long time.

**Comment by TimOliver** [comment]
@jpsim Fair enough!

I refactored the export code to be handled on a separate queue while displaying this dialog:
realmexportindicator

Let me know if you think that needs re-wording, or anything else! :)

File: RealmBrowser/Classes/RLMRealmBrowserWindowController.m [hunk 3]
```diff
@@ -101,6 +108,80 @@ - (IBAction)saveObjcModels:(id)sender
...
+        alert.messageText = @"You cannot overwrite the original Realm file.";
+        alert.informativeText = @"Please choose a different location in which to save this Realm file.";
+        [alert runModal];
+        return;
+    }
+    
+    //Ensure a file with the same name doesn't already exist
+    BOOL directory = NO;
+    
+    if ([[NSFileManager defaultManager] fileExistsAtPath:realmFileURL.path isDirectory:&directory] && !directory) {
```
Comment (jpsim) [inline RealmBrowser/Classes/RLMRealmBrowserWindowController.m]: isn't it common to prompt before overwriting a file? I don't know if NSSavePanel would ever return a directory here, but if it does, this code will skip this deletion step and attempt to write a realm file to a location matching a directory, which will throw an exception.

**Comment by jpsim** [comment]
I think that progress indicator should help, so :+1: on that. I just have the one last comment about overwriting files without prompt, and potentially attempting to write the new realm file over an existing path containing a directory.

File: RealmBrowser/Classes/RLMRealmBrowserWindowController.m (hunk 3 above)
Comment (TimOliver) [inline RealmBrowser/Classes/RLMRealmBrowserWindowController.m]: It does, but I discovered in testing that if you try and save a Realm file over itself while it's open in Browser, it'll throw an exception.

I had a look through Apple's documentation. `NSSavePanel` does *not* allow you to select a directory, it must explicitly be a file (However `NSOpenPanel` apparently does let you select a directory), so this logic should work okay as-is.

**Pull request merged**